app/peptide/payloadstore/payloadstore.go

```diff
@@ -1,70 +1,100 @@
 package payloadstore
 
 import (
-	"sync"
+	"fmt"
 
 	"github.com/ethereum/go-ethereum/beacon/engine"
 	"github.com/polymerdao/monomer"
 )
 
-type PayloadStore interface {
-	Add(payload *monomer.Payload)
-	Get(id engine.PayloadID) (*monomer.Payload, bool)
-	Current() *monomer.Payload
-	RollbackToHeight(height int64) error
-}
-
-type pstore struct {
-	mutex    sync.Mutex
+type Store struct {
 	payloads map[engine.PayloadID]*monomer.Payload
 	heights  map[int64]engine.PayloadID
 	current  *monomer.Payload
 }
 
-var _ PayloadStore = (*pstore)(nil)
-
-func NewPayloadStore() PayloadStore {
-	return &pstore{
-		mutex:    sync.Mutex{},
+func NewPayloadStore() *Store {
+	return &Store{
 		payloads: make(map[engine.PayloadID]*monomer.Payload),
 		heights:  make(map[int64]engine.PayloadID),
 	}
 }
 
-func (p *pstore) Add(payload *monomer.Payload) {
+// Add adds a payload to the store and updates the current payload
+// if the new payload has a greater height.
+func (p *Store) Add(payload *monomer.Payload) {
 	id := payload.ID()
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
 	if _, ok := p.payloads[*id]; !ok {
 		p.heights[payload.Height] = *id
 		p.payloads[*id] = payload
-		p.current = payload
+
+		if (p.current == nil) || (payload.Height > p.current.Height) {
+			p.current = payload
+		}
 	}
 }
 
-func (p *pstore) Get(id engine.PayloadID) (*monomer.Payload, bool) {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
+// Remove removes a payload from the store and updates the current payload
+// if necessary.
+func (p *Store) Remove(id engine.PayloadID) {
+	current := p.Current()
+	resetCurrent := id == *current.ID()
+
 	if payload, ok := p.payloads[id]; ok {
-		return payload, true
+		delete(p.payloads, id)
+		delete(p.heights, payload.Height)
+	}
+
+	if resetCurrent {
+		p.current = nil
+		for _, payload := range p.payloads {
+			if p.current == nil || payload.Height > p.current.Height {
+				p.current = payload
+			}
+		}
 	}
-	return nil, false
 }
 
-func (p *pstore) Current() *monomer.Payload {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
+func (p *Store) Get(id engine.PayloadID) (*monomer.Payload, bool) {
+	payload, ok := p.payloads[id]
+	return payload, ok
+}
+
+func (p *Store) Current() *monomer.Payload {
 	return p.current
 }
 
-func (p *pstore) RollbackToHeight(height int64) error {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
+// RollbackToHeight removes all payloads at heights greater than the given height
+// and sets the current payload to the highest height payload that remains.
+//
+// If there is no payload at the given height, it still performs the pruning,
+// but returns an error as an indication.
+func (p *Store) RollbackToHeight(height int64) error {
+	max := int64(0)
 
+	for h, id := range p.heights {
+		if h > height {
+			delete(p.payloads, id)
+			delete(p.heights, h)
+		} else if h > max {
+			max = h
+		}
+	}
+
+	// set the current payload to the highest height
+	p.current = p.payloads[p.heights[max]]
```

Comment:
Similar to above. The current payload is the "next block", not the "highest block."